src/primaite/agents/git/git_policy.py

```diff
@@ -1,28 +1,36 @@
 import torch
 import torch.nn.functional as F
 from torch import nn
-from torch_geometric.nn import GATConv
-from torch_geometric.nn.pool import global_max_pool
+from torch_geometric.nn import GATConv, LayerNorm, global_add_pool
 from torch.optim import Adam
 from transformers import BertModel, BertTokenizer, BitsAndBytesConfig
 from transformers import AutoModelForCausalLM, AutoTokenizer
 from transformers.models.llama.modeling_llama import LlamaForCausalLM
 from typing import List, Tuple, Dict
 import numpy as np
+import logging
+
+
+logging.getLogger().setLevel(logging.INFO)
 
 
 class GraphEmbedding(nn.Module):
-    def __init__(self, in_channels, output_dim, hidden_dim, device: str = "cuda:1"):
+    def __init__(self, in_channels, output_dim, hidden_dim, n_tokens: int = 10, device: str = "cuda:1"):
         super().__init__()
         self.gat = GATConv(in_channels=in_channels, out_channels=hidden_dim)
-        self.linear = nn.Linear(in_features=hidden_dim, out_features=output_dim)
+        self.layer_norm = LayerNorm(hidden_dim)
+        self.linear = nn.Linear(in_features=hidden_dim, out_features=n_tokens * output_dim)
         self.device = device
+        self.n_tokens = n_tokens
+        self.output_dim = output_dim
 
     def forward(self, x, edge_index):
+        n_nodes = x.shape[0]
         x = self.gat(x.to(self.device), edge_index.to(self.device))
```

Comment:
dw about this for now, but for future reference it would be good to have several layers of GAT

also, every time you apply a convolution layer, you should follow it with a layer_norm application and then a relu application

Reply:
this is because GAT will only apply one round of message passing so if you want to capture further distances in the graph you need multiple rounds

better check the gnn notebooks for reference

again dw about this for now but keep this in mind

but be aware that if you have multiple gan layers in a row, that you add a jumping knowledge layer at the end: https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.JumpingKnowledge.html

check notebook 4 from the graph representation learning practicals, there's a really good example there for how to do it!

Reply:
you need to do this with multiple gat layers because otherwise the vectors get "oversmoothed"

Reply:
I made my knowledge jump prrr